### PR TITLE
[Etags] Return Koala::HTTPService::Response.new 

### PR DESCRIPTION
**Pull request by jessieay**
- when `http_component: :response` option set
- Can use that option to return response with status, body, headers
- Set body to either parsed JSON or GraphCollection object
- Need to know response status to determine if an etag header passed in matches (304)
- Reference: https://developers.facebook.com/docs/marketing-api/etags
- Fixes arsduo#337

### CODE REVIEW

**Comment by jessieay** [comment]
@arsduo would love your feedback on this. I am already actively using my fork on a project that depends heavily on etag caching and so far so good!

**Comment by arsduo** [comment]
Hey Jessie! Thanks for the pull request, and sorry it's taken me so long to review it! I haven't used the APIs with etags, but they seem awesome.

I need to read the code over again in more detail later this weekend — it's changing a few key points, so I worry that it could break backward compatibility. (That might not be a bad thing, given that the raw response should perhaps try to parse the JSON. Or maybe it should stay raw. I'll think about it.)

I wonder if we could simplify this in a way that more accurately matches what we want to do here. It seems like we want the following:

* to include the etag with the outgoing request
* to get back a response that indicates whether we got a 304 – Not Modified back
* if we don't get a 304 back, we want both the response data and the etag

If that's correct, what do you think of the following approach? 

* accept an :etag parameter In the http_options hash
* if that value is present, the etag value is sent automatically in the header
* if that value is present, the response is returned not as a string or Response object, but as anew EtagResponse object, which would expose `data_has_changed?`, `result`, and `new_etag` (or whatever names make the most sense). If there's other info thats relevant, we could include that too.

If I understand the need correctly, building something like this might require fewer changes to the existing functionality, and to give etag users easier access to their data.

If this approach makes sense, do you want to try implementing it? If not, I can. 

(I'll be releasing 2.0 soon, which refactors some code you'd need to touch, so if you want to work on this, you should base the code on the 2.0 branch. If that changes tomorrow when I do a bunch of merging, I'll let you know,)

Thanks and have a great weekend!

**Comment by jessieay** [comment]
Hi @arsduo - thanks for the detailed reply.

I agree that we should not break backward compatibility. Since we are taking advantage of `http_component: response` here, though, I was thinking that it does not actually break backward compatibility, since passing in that option returned a `response` prior to this change - it . 

The only difference is that we are now properly parsing the response body of a request with the `http_component: :response` option. Before, there was a parsing error at https://github.com/jessieay/koala/blob/master/lib/koala/api/graph_api.rb#L498 when using this option.

I can include more detailed info if you like. I like the explicitness of your solution for etags, but I also like the flexibility of just passing in `http_component: :response` when making a request where you want response status and headers.

**Comment by jessieay** [comment]
Hi @arsduo ! 

Upon revisiting this and playing around with my code that uses these changes vs what's on master, I think the only thing this PR is actually changing is that graph calls that use the `http_component: :response` option are now returning a parsed response body rather than a string for the response body. 

As you can see, using the `http_component: :response` option was returning an instance of `Koala::HTTPService::Response` before: https://github.com/arsduo/koala/pull/454/files#diff-efe9ef87d71041c98170caaa8d38ef41L39 (despite the title of this PR, which implies that this was not happening before)

With this change, we are parsing the response body in the same way the response body is parsed for graph calls that do *not* use the `http: :response` option. See https://github.com/arsduo/koala/blob/master/lib/koala/api.rb#L82 for reference. 

The only difference is that the parsed body is being set as the `body` attribute for a ` Koala::HTTPService::Response` instance rather than returning only the parsed body.

As background, I ran into this issue because I wanted to return the HTTP status code for my requests because I was using etags to check for cached responses from Facebook. When I got back responses that were not cached and went to parse the body, it felt off to have to parse the response body differently than I was parsing it for calls that did not use the `http_component: :response` option.

I think the approach you highlighted above is fine, but perhaps too inflexible. If the Facebook API starts using caching with something other than etags, that would required a change to this library. I think that passing in the `http_component: :response` is a nice, flexible way of asking for request headers and HTTP response code rather than just a response body. I imagine there are many current and possible future reasons why someone would want this information. But returning the response body as a string is unexpected.

In conclusion, you are right - this *does* change existing functionality. If someone is currently using the `http_component: :response` option in their graph API calls and handling the parsing of the response body themselves, this would prove problematic. So perhaps best to include in the next release. But I do think the current implementation is flexible and would rather not go down the road of creating an object specifically for etags. 

Thoughts?

**Comment by jessieay** [comment]
Hey there - just a nudge to @arsduo - any thoughts on my last comment?

**Comment by sailing** [comment]
@arsduo @jessieay this is becoming more and more relevant, especially following FB's new rate limiting protocols for apps. https://developers.facebook.com/docs/graph-api/advanced/rate-limiting 

Thoughts on moving this forward?

**Comment by jessieay** [comment]
@sailing thanks for the bump, had forgotten this was still open. 

I've been using a fork with these changes on a production application and it has been working. It does introduce a breaking change, though, so would want it to be released as part of a version update.

**Comment by jessieay** [comment]
Hello @arsduo I see this now has a merge conflict. Thoughts on how we can move it forward?

**Comment by arsduo** [comment]
I totally dropped the ball on this. I'll review again sometime soon.

File: lib/koala/api.rb [hunk 1]
```diff
@@ -77,7 +77,13 @@ def api(path, args = {}, verb = "get", options = {}, &error_checking_block)
 
         # if we want a component other than the body (e.g. redirect header for images), return that
         if component = options[:http_component]
-          component == :response ? result : result.send(options[:http_component])
+          if component == :response
+            parsed_body = MultiJson.load("[#{result.body.to_s}]")[0]
```
Comment (arsduo) [inline lib/koala/api.rb]: Koala's no longer using MultiJson -- there's an example JSON.parse line that you can copy below in the updated version.

**Comment by arsduo** [comment]
After thinking about this some more and looking at the code, I realize this is simpler than I thought. 

I'm removing the REST API in version 3, since it's no longer supported, and that means that this raw API call is only used by the Graph API. Given that, I ultimately want to merge them together, but for now we can change the interface to return a result object, which can be incorporated into the GraphCollection.

Since this is a little bit more surgery, I'll take it on this afternoon using the code you've already written. Should have this done soon.

**Comment by arsduo** [comment]
#589 incorporates similar changes against v3.0, so I'll close this. Thank you for submitting this PR and apologies again that it took so long to respond. I'll be crediting you as well for this change in the changelog.

**Comment by jessieay** [comment]
Hi @arsduo ! Thanks for opening the new PR. No worries on the delay, I know how it can be maintaining open source libs :) ❤️
